src/rpc/client.zig

```diff
@@ -1,2338 +1,537 @@
-        req.transfer_encoding = .chunked;
-        defer req.deinit();
-        try req.start();
-
-        const JrpcRequest = jsonrpc.Request(Params);
-
-        const out = try std.json.stringifyAlloc(self.allocator, JrpcRequest{
-            .jsonrpc = "2.0",
-            .id = id,
-            .method = method,
-            .params = params,
-        }, .{ .emit_null_optional_fields = false });
-        defer self.allocator.free(out);
-        logger.debug("request body: {s}", .{out});
-
-        try req.writer().writeAll(out);
-        try req.finish();
-        logger.debug("request wating..", .{});
-        try req.wait();
-
-        if (req.response.status != http.Status.ok) {
-            logger.warn("response status not 200: {}", .{req.response.status});
-            return Error.ResponseNotStatusOk;
-        }
-
-        logger.debug("request reading..", .{});
-        const body = try req.reader().readAllAlloc(self.allocator, 819200);
-        defer self.allocator.free(body);
-
-        logger.debug("response body: {s}", .{body});
-
-        var parser = std.json.Parser.init(self.allocator, .alloc_always);
-        defer parser.deinit();
-
-        const tree = try parser.parse(body);
-        return tree;
-    }
-
-    pub fn getSlot(self: *Self) !jsonrpc.Response(i64) {
-        return try self.makeRequestWithJRpcResponse(i64, ?u8, "getSlot", "1", null);
-    }
-
-    const GetAccountInfoOptions = struct {
-        commitment: ?types.Commitment = null,
-        encoding: types.Encoding = .Base64,
-    };
-
-    pub fn getAccountInfo(self: *Self, pubkey: Pubkey, options: GetAccountInfoOptions) !jsonrpc.Response(types.AccountInfo) {
-        var optionsObj = std.StringArrayHashMap(json.Value).init(self.allocator);
-        defer optionsObj.deinit();
-        try optionsObj.put("encoding", .{ .string = options.encoding.string() });
-        try optionsObj.put("commitment", .{ .string = self.defaultCommitmentOr(options.commitment).string() });
-
-        var arrList = std.ArrayList(json.Value).init(self.allocator);
-        defer arrList.deinit();
-
-        const str = pubkey.string();
-
-        try arrList.append(.{ .string = str });
-        try arrList.append(.{ .object = optionsObj });
-
-        const params = .{ .array = arrList };
-
-        const resp = try self.makeRequestWithJRpcResponse(types.AccountInfo, json.Value, "getAccountInfo", "1", params);
-        return resp;
+    pub fn deinit(self: *Client) void {
+        self.client.deinit();
     }
 
-    pub fn getBalance(self: *Self, pubkey: Pubkey) !jsonrpc.Response(types.BalanceInfo) {
-        const str = pubkey.string();
+    /// Wraps a parsed response from the RPC server with an arena
+    /// used for request, response, and json parsing allocations
+    pub fn Result(comptime T: type) type {
+        return struct {
+            arena: std.heap.ArenaAllocator,
+            value: T,
```

Comment:
Just to be sure and have this in writing, is there any instance where a `T` is a slightly more complex data structure, or contains one, ie `std.ArrayList`? If there is, this being a stack copy is somewhat dangerous, since the reference to the allocator held by something like a managed arraylist would become invalid after the function returns this.
If there are no cases where that's an issue, just slap a comment on this about that and we're good - if there are, we should probably mimic modern `std.json.Parsed` and allocate stable space for it on the heap, turning this into `arena: std.heap.ArenaAllocator`.

Reply:
Good catch, 0ef63e8